### PR TITLE
Feature/python wrapper

### PR DESCRIPTION
**Pull request by grahamrow**
@caryan @blakejohnson Just created a crude python wrapper for libaps2. Could be turned into a class with some nice decorators for the redundant parts, but I wanted to get input first. Then again, most of the wrapping could take place in pycontrol.

### CODE REVIEW

File: src/python/aps.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
+import numpy as np
+import numpy.ctypeslib as npct
+from ctypes import c_int, c_uint, c_ulong, c_ulonglong, c_float, c_double, c_char, c_char_p, addressof, create_string_buffer, byref, POINTER
+
+libaps2 = npct.load_library("libaps2", ".")
```
Comment (caryan) [inline src/python/aps.py]: Let's assume library is in `build` for now. Also have you tested dependency loading on Windows?

File: src/python/aps.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
+import numpy as np
+import numpy.ctypeslib as npct
```
Comment (caryan) [inline src/python/aps.py]: Why numpy ctypes?

File: src/python/aps.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+np_double_1D = npct.ndpointer(dtype=np.double,  ndim=1, flags='CONTIGUOUS')
+np_float_1D  = npct.ndpointer(dtype=np.float,   ndim=1, flags='CONTIGUOUS')
+np_int16_1D  = npct.ndpointer(dtype=np.int16,   ndim=1, flags='CONTIGUOUS')
+np_int8_1D   = npct.ndpointer(dtype=np.int8,    ndim=1, flags='CONTIGUOUS')
+np_uint64_1D = npct.ndpointer(dtype=np.uint64,  ndim=1, flags='CONTIGUOUS')
+np_uint32_1D = npct.ndpointer(dtype=np.uint32,  ndim=1, flags='CONTIGUOUS')
+np_uint16_1D = npct.ndpointer(dtype=np.uint16,  ndim=1, flags='CONTIGUOUS')
+np_uint8_1D  = npct.ndpointer(dtype=np.uint8,   ndim=1, flags='CONTIGUOUS')
+
+class Flag(object):
```
Comment (caryan) [inline src/python/aps.py]: What's this class for?

File: src/python/aps.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    flags = [(0x1, 'fun'), (0x2, 'toy')]
+    @classmethod
+    def from_param(cls, data):
+        return c_uint(encode_flags(self.flags, data))
+
+# APS2_TRIGGER_SOURCE
```
Comment (caryan) [inline src/python/aps.py]: It's really unfortunate there is no automatic way to do this.

**Comment by caryan** [comment]
It's a functional start but why don't we wrap everything in an `ap2_call` and `aps2_getter` like we do in Julia and Matlab?

File: src/python/aps.py (hunk 2 above)
Comment (grahamrow) [inline src/python/aps.py]: So we can pass numpy arrays directly to waveform/sequence/marker functions.

File: src/python/aps.py (hunk 1 above)
Comment (grahamrow) [inline src/python/aps.py]: Windows remains untested. Should we do a relative import? ```../build```?

File: src/python/aps.py (hunk 3 above)
Comment (grahamrow) [inline src/python/aps.py]: Leftover cruft from the inevitable stack overflow binge...

File: src/python/aps.py (hunk 4 above)
Comment (grahamrow) [inline src/python/aps.py]: Blake was saying we might look into cffi, though I'm not sure how that matches up yet. Multiple cursors did about 80% of the work, luckily. Swig doesn't actually seem to make things any easier in this case.

Comment (caryan) [inline src/python/aps.py]: I thought you needed a C compiler for cffi which might be a drag for deployment (although of course Matlab famously does too)

File: src/python/aps.py (hunk 1 above)
Comment (caryan) [inline src/python/aps.py]: Seems to depend on OS and language, especially for mingw dependencies.  But we can try just relative as see if it works.

File: src/python/aps.py (hunk 4 above)
Comment (rmcgurrin) [inline src/python/aps.py]: It's worth checking out libAlazar.py in libAlazar for consistency sake in how the libs are handles.

Comment (blakejohnson) [inline src/python/aps.py]: I just noticed that numba could automatically translate cffi calls when JITing, but I don't know yet what are cffi's requirements.

**Comment by grahamrow** [comment]
Moved to a class/metaclass approach in d9c6ae0b20ad8530b3fe46b264c1dfc68fe48364. The enums aren't working yet, and I get crazy integers for the time being.

**Comment by caryan** [comment]
I think it's for better

**Comment by caryan** [comment]
Can we rename to aps2.py for consistency.

**Comment by grahamrow** [comment]
What looks strange, the ```getattr(libaps2, name).var = something``` business? Maybe some ```eval``` calls instead? 😜

**Pull request merged**